src/workerd/jsg/jsvalue.h

```diff
@@ -0,0 +1,467 @@
+#pragma once
+
+#include "jsg.h"
+#include <v8.h>
+
+namespace workerd::jsg {
+
+inline void requireOnStack(void* self) {
+#ifdef KJ_DEBUG
+  kj::requireOnStack(self, "JsValue types must be allocated on stack");
+#endif
+}
+
+// The types listed in the JS_IS_TYPES macro are translated into is{Name}()
+// methods on the JsValue type. These correspond directly to equivalent v8::Value
+// types and therefore must be kept in sync.
+#define JS_IS_TYPES(V) \
+  V(Undefined) \
+  V(Null) \
+  V(NullOrUndefined) \
+  V(True) \
+  V(False) \
+  V(ArgumentsObject) \
+  V(NativeError) \
+  V(Name) \
+  V(Function) \
+  V(AsyncFunction) \
+  V(GeneratorFunction) \
+  V(GeneratorObject) \
+  V(WeakMap) \
+  V(WeakSet) \
+  V(WeakRef) \
+  V(WasmNull) \
+  V(ModuleNamespaceObject) \
+  V(MapIterator) \
+  V(SetIterator) \
+  V(External) \
+  V(BigIntObject) \
+  V(BooleanObject) \
+  V(NumberObject) \
+  V(StringObject) \
+  V(SymbolObject) \
+  V(Promise) \
+  V(ArrayBuffer) \
+  V(ArrayBufferView) \
+  V(TypedArray) \
+  V(Uint8Array) \
+  V(Uint8ClampedArray) \
+  V(Int8Array) \
+  V(Uint16Array) \
+  V(Int16Array) \
+  V(Uint32Array) \
+  V(Int32Array) \
+  V(Float32Array) \
+  V(Float64Array) \
+  V(BigInt64Array) \
+  V(BigUint64Array) \
+  V(DataView) \
+  V(Proxy) \
+  V(SharedArrayBuffer) \
+  V(WasmMemoryObject) \
+  V(WasmModuleObject) \
+  JS_TYPE_CLASSES(V)
+
+template <typename TypeWrapper>
+struct JsValueWrapper;
+
+// A JsValue is an abstraction for a JavaScript value that has not been mapped
```

Comment:
A comparison of `jsg::Value` and `jsg::JsValue` use cases would probably be useful.